OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs

```diff
@@ -72,6 +72,18 @@ void INotifyCreated.Created(Actor self)
 				player.PlayerActor.Trait<FrozenActorLayer>().Add(frozenActor);
 				return new FrozenState(frozenActor) { IsVisible = startsRevealed };
 			});
+
+			// Set the initial visibility state
+			// This relies on actor.GetTargetablePositions(), which is also setup up in Created.
+			// Since we can't be sure whether our method will run after theirs, defer by a frame.
+			if (startsRevealed)
+				self.World.AddFrameEndTask(_ =>
+				{
+					for (var playerIndex = 0; playerIndex < frozenStates.Count; playerIndex++)
+						UpdateFrozenActor(frozenStates[playerIndex].FrozenActor, playerIndex);
+				});
+
+			created = true;
```

Comment:
In the old code this was only set if `startsRevealed` is `true` and was set at the same time the update ran, so should be inside the FrameEndTask as well?

Reply:
The new logic that's been added to `ICreatesFrozenActors.OnVisibilityChanged` needs to run in both cases. Trying out your suggestion on a map with Explored Map = false results in buildings being invisible. They are invisible both if you have sight and if you do not.